File: JokeGenerator.cs
// JokeGenerator.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;

class Joke
{
    public int Id { get; set; }
    public string Text { get; set; }
    public string Category { get; set; }
}

class JokeGenerator
{
    private List<Joke> jokes = new List<Joke>();
    private Random rand = new Random();

    public JokeGenerator()
    {
        LoadDefaultJokes();
    }

    private void LoadDefaultJokes()
    {
        jokes = new List<Joke>
        {
            new Joke { Id = 1, Text = "Why don't scientists trust atoms? Because they make up everything!", Category = "science" },
            new Joke { Id = 2, Text = "What do you call a fake noodle? An impasta!", Category = "food" },
            new Joke { Id = 3, Text = "Why did the scarecrow win an award? Because he was outstanding in his field!", Category = "pun" },
            new Joke { Id = 4, Text = "I'm reading a book on anti-gravity. It's impossible to put down!", Category = "book" },
            new Joke { Id = 5, Text = "What's the best thing about Switzerland? I don't know, but the flag is a big plus!", Category = "geography" },
            new Joke { Id = 6, Text = "Why do programmers prefer dark mode? Because light attracts bugs!", Category = "programming" },
            new Joke { Id = 7, Text = "I told my wife she was drawing her eyebrows too high. She looked surprised.", Category = "marriage" },
            new Joke { Id = 8, Text = "What do you call a bear with no teeth? A gummy bear!", Category = "animal" },
            new Joke { Id = 9, Text = "Why don't eggs tell jokes? They'd crack each other up!", Category = "food" },
            new Joke { Id = 10, Text = "I used to play piano by ear, but now I use my hands.", Category = "music" },
        };
    }

    public Joke GetRandomJoke()
    {
        if (jokes.Count == 0) return null;
        return jokes[rand.Next(jokes.Count)];
    }

    public List<Joke> GetAllJokes() => jokes;

    public Joke AddJoke(string text, string category = "general")
    {
        int id = jokes.Count > 0 ? jokes.Max(j => j.Id) + 1 : 1;
        var joke = new Joke { Id = id, Text = text, Category = category };
        jokes.Add(joke);
        return joke;
    }

    public bool RemoveJoke(int id)
    {
        var joke = jokes.FirstOrDefault(j => j.Id == id);
        if (joke != null) return jokes.Remove(joke);
        return false;
    }

    public List<Joke> GetByCategory(string category)
    {
        return jokes.Where(j => j.Category.Equals(category, StringComparison.OrdinalIgnoreCase)).ToList();
    }

    public List<string> GetCategories()
    {
        return jokes.Select(j => j.Category).Distinct().OrderBy(c => c).ToList();
    }

    public void SaveToFile(string filename)
    {
        string json = JsonSerializer.Serialize(jokes, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(filename, json);
    }

    public void LoadFromFile(string filename)
    {
        if (!File.Exists(filename)) return;
        string json = File.ReadAllText(filename);
        jokes = JsonSerializer.Deserialize<List<Joke>>(json) ?? new List<Joke>();
    }

    public static void DisplayJoke(Joke joke)
    {
        if (joke == null)
        {
            Console.WriteLine("No jokes available.");
            return;
        }
        Console.WriteLine($"\n😂 Joke: {joke.Text}");
        if (!string.IsNullOrEmpty(joke.Category))
            Console.WriteLine($"📁 Category: {joke.Category}");
    }

    static void Main()
    {
        var gen = new JokeGenerator();
        Console.WriteLine("=== Random Joke Generator ===");
        while (true)
        {
            Console.WriteLine("\n1. Get a random joke");
            Console.WriteLine("2. Show all jokes");
            Console.WriteLine("3. Add a joke");
            Console.WriteLine("4. Remove a joke");
            Console.WriteLine("5. Filter jokes by category");
            Console.WriteLine("6. Show categories");
            Console.WriteLine("7. Save jokes to file");
            Console.WriteLine("8. Load jokes from file");
            Console.WriteLine("9. Exit");
            Console.Write("Choose: ");
            string choice = Console.ReadLine()?.Trim() ?? "";
            switch (choice)
            {
                case "1":
                    DisplayJoke(gen.GetRandomJoke());
                    break;
                case "2":
                    var all = gen.GetAllJokes();
                    if (all.Count == 0) Console.WriteLine("No jokes.");
                    else
                    {
                        Console.WriteLine("\nAll jokes:");
                        foreach (var j in all)
                            Console.WriteLine($"[{j.Id}] {j.Text} ({j.Category})");
                    }
                    break;
                case "3":
                    Console.Write("Enter your joke: ");
                    string text = Console.ReadLine()?.Trim();
                    if (string.IsNullOrEmpty(text)) { Console.WriteLine("Joke cannot be empty."); break; }
                    Console.Write("Category (optional): ");
                    string cat = Console.ReadLine()?.Trim() ?? "general";
                    var joke = gen.AddJoke(text, cat);
                    Console.WriteLine($"Joke added with ID {joke.Id}.");
                    break;
                case "4":
                    Console.Write("Enter joke ID to remove: ");
                    if (int.TryParse(Console.ReadLine(), out int id))
                    {
                        if (gen.RemoveJoke(id)) Console.WriteLine("Joke removed.");
                        else Console.WriteLine("Joke not found.");
                    }
                    else Console.WriteLine("Invalid ID.");
                    break;
                case "5":
                    Console.Write("Enter category: ");
                    string catFilter = Console.ReadLine()?.Trim() ?? "";
                    var filtered = gen.GetByCategory(catFilter);
                    if (filtered.Count == 0)
                        Console.WriteLine($"No jokes in category '{catFilter}'.");
                    else
                    {
                        Console.WriteLine($"\nJokes in category '{catFilter}':");
                        foreach (var j in filtered)
                            Console.WriteLine($"[{j.Id}] {j.Text}");
                    }
                    break;
                case "6":
                    var cats = gen.GetCategories();
                    if (cats.Count == 0) Console.WriteLine("No categories.");
                    else Console.WriteLine("Categories: " + string.Join(", ", cats));
                    break;
                case "7":
                    Console.Write("Filename: ");
                    string fname = Console.ReadLine()?.Trim() ?? "jokes.json";
                    gen.SaveToFile(fname);
                    Console.WriteLine($"Saved to {fname}.");
                    break;
                case "8":
                    Console.Write("Filename: ");
                    fname = Console.ReadLine()?.Trim() ?? "jokes.json";
                    gen.LoadFromFile(fname);
                    Console.WriteLine($"Loaded from {fname}.");
                    break;
                case "9":
                    Console.WriteLine("Goodbye! 😄");
                    return;
                default:
                    Console.WriteLine("Invalid choice.");
                    break;
            }
        }
    }
}

File: joke_generator.rb
# joke_generator.rb
require 'json'

class JokeGenerator
  def initialize
    @jokes = []
    load_default_jokes
  end

  def load_default_jokes
    @jokes = [
      { id: 1, text: "Why don't scientists trust atoms? Because they make up everything!", category: "science" },
      { id: 2, text: "What do you call a fake noodle? An impasta!", category: "food" },
      { id: 3, text: "Why did the scarecrow win an award? Because he was outstanding in his field!", category: "pun" },
      { id: 4, text: "I'm reading a book on anti-gravity. It's impossible to put down!", category: "book" },
      { id: 5, text: "What's the best thing about Switzerland? I don't know, but the flag is a big plus!", category: "geography" },
      { id: 6, text: "Why do programmers prefer dark mode? Because light attracts bugs!", category: "programming" },
      { id: 7, text: "I told my wife she was drawing her eyebrows too high. She looked surprised.", category: "marriage" },
      { id: 8, text: "What do you call a bear with no teeth? A gummy bear!", category: "animal" },
      { id: 9, text: "Why don't eggs tell jokes? They'd crack each other up!", category: "food" },
      { id: 10, text: "I used to play piano by ear, but now I use my hands.", category: "music" },
    ]
  end

  def random_joke
    @jokes.sample
  end

  def all_jokes
    @jokes
  end

  def add_joke(text, category = 'general')
    id = @jokes.map { |j| j[:id] }.max.to_i + 1
    joke = { id: id, text: text, category: category }
    @jokes << joke
    joke
  end

  def remove_joke(id)
    @jokes.reject! { |j| j[:id] == id }
  end

  def by_category(category)
    @jokes.select { |j| j[:category].downcase == category.downcase }
  end

  def categories
    @jokes.map { |j| j[:category] }.uniq.sort
  end

  def save_to_file(filename)
    File.write(filename, JSON.pretty_generate(@jokes))
  end

  def load_from_file(filename)
    data = File.read(filename)
    @jokes = JSON.parse(data, symbolize_names: true)
  rescue Errno::ENOENT
    puts "File not found."
  end

  def display_joke(joke)
    if joke.nil?
      puts "No jokes available."
      return
    end
    puts "\n😂 Joke: #{joke[:text]}"
    puts "📁 Category: #{joke[:category]}" if joke[:category]
  end
end

def main
  gen = JokeGenerator.new
  puts "=== Random Joke Generator ==="
  loop do
    puts "\n1. Get a random joke"
    puts "2. Show all jokes"
    puts "3. Add a joke"
    puts "4. Remove a joke"
    puts "5. Filter jokes by category"
    puts "6. Show categories"
    puts "7. Save jokes to file"
    puts "8. Load jokes from file"
    puts "9. Exit"
    print "Choose: "
    choice = gets.chomp.strip
    case choice
    when '1'
      gen.display_joke(gen.random_joke)
    when '2'
      jokes = gen.all_jokes
      if jokes.empty?
        puts "No jokes."
      else
        puts "\nAll jokes:"
        jokes.each { |j| puts "[#{j[:id]}] #{j[:text]} (#{j[:category]})" }
      end
    when '3'
      print "Enter your joke: "
      text = gets.chomp.strip
      if text.empty?
        puts "Joke cannot be empty."
        next
      end
      print "Category (optional): "
      cat = gets.chomp.strip
      cat = 'general' if cat.empty?
      joke = gen.add_joke(text, cat)
      puts "Joke added with ID #{joke[:id]}."
    when '4'
      print "Enter joke ID to remove: "
      id = gets.chomp.to_i
      if gen.remove_joke(id)
        puts "Joke removed."
      else
        puts "Joke not found."
      end
    when '5'
      print "Enter category: "
      cat = gets.chomp.strip
      jokes = gen.by_category(cat)
      if jokes.empty?
        puts "No jokes in category '#{cat}'."
      else
        puts "\nJokes in category '#{cat}':"
        jokes.each { |j| puts "[#{j[:id]}] #{j[:text]}" }
      end
    when '6'
      cats = gen.categories
      if cats.empty?
        puts "No categories."
      else
        puts "Categories: #{cats.join(', ')}"
      end
    when '7'
      print "Filename: "
      fname = gets.chomp.strip
      gen.save_to_file(fname)
      puts "Saved to #{fname}."
    when '8'
      print "Filename: "
      fname = gets.chomp.strip
      gen.load_from_file(fname)
      puts "Loaded from #{fname}."
    when '9'
      puts "Goodbye! 😄"
      break
    else
      puts "Invalid choice."
    end
  end
end

main if __FILE__ == $0
